# MacSGP: Mapping Cell-Type-Specific Spatial Gene Programs Uncovers Tissue Architecture and Microenvironment Organization

<!-- badges: start -->
[![pypi](https://img.shields.io/pypi/v/MacSGP.svg)](https://pypi.python.org/pypi/MacSGP/)
![GitHub repo
size](https://img.shields.io/github/repo-size/YangLabHKUST/MacSGP)
![GitHub last
commit](https://img.shields.io/github/last-commit/YangLabHKUST/MacSGP)
![GitHub
License](https://img.shields.io/github/license/YangLabHKUST/MacSGP)
[![HitCount](https://hits.dwyl.com/YangLabHKUST/MacSGP.svg?style=flat-square)](http://hits.dwyl.com/YangLabHKUST/MacSGP)
[![PyPI Downloads](https://static.pepy.tech/personalized-badge/macsgp?period=total&units=INTERNATIONAL_SYSTEM&left_color=BLACK&right_color=GREEN&left_text=downloads)](https://pepy.tech/projects/macsgp)
![GitHub Repo
stars](https://img.shields.io/github/stars/YangLabHKUST/MacSGP)
![GitHub
forks](https://img.shields.io/github/forks/YangLabHKUST/MacSGP)

<!-- badges: end -->

## Introduction

MacSGP is a scalable statistical and computational approach for MApping Cell-type-specific Spatial Gene Programs (SGPs) in spatial transcriptomic (ST) data.

MacSGP's effectiveness relies on our innovations in the seamless integration of deep graph neural networks (GNNs) and probabilistic models:

- MacSGP maps gene expressions and spatial information of spots into a shared latent space by leveraging deep GNNs, yielding low-dimensional representations of each spot that capture both gene expression similarity and spatial coherence.
- MacSGP utilizes the latent representation to generate cell-type-specific SGPs through a probabilistic model, which accounts for cell type mixtures and characterizes cell-type-specific SGPs using the low-rank structure.
- For large-scale high-resolution ST datasets, MacSGP adopts a batch-learning scheme that learns SGPs over small gene patches, enabling scalable training without sacrificing accuracy. 

<figure>

<img src="figures/overview.png" style="width:95.0%"
alt="Overview" />
</figure>

## Installation
It's recommended to create a virtual environment first.

```shell
$ conda create -n MacSGP python=3.11
$ conda activate MacSGP
```

### Requirements
MacSGP requires `pytorch` and `PyG` 

[Install Pytorch](https://pytorch.org/get-started/locally/) [Install PyG](https://pytorch-geometric.readthedocs.io/en/latest/install/installation.html)

For `PyG`, MacSGP also requires its additional libraries, their installation requires specifications for torch version and CUDA version. Users could use `nvcc --version` to check the CUDA version for installation.

Here we provide with an example of the CUDA `12.8` installation code.

```shell
$ pip install torch_geometric

# Additional dependencies:
$ pip install pyg_lib torch_scatter torch_sparse torch_cluster torch_spline_conv -f https://data.pyg.org/whl/torch-2.8.0+cu128.html
```

### Installation through PyPI
MacSGP can be installed from PyPI:

```shell
$ pip install MacSGP
```

### Installation through GitHub
Alternatively, MacSGP can be downloaded from GitHub:
```shell
# Clone the repository
$ git clone https://github.com/YangLabHKUST/MacSGP.git
$ cd MacSGP

# Install the required packages
$ pip install -r requirements.txt

# Install MacSGP
$ python setup.py build
$ python setup.py install
```

## Real data analysis

The code for reproducing the real data analysis results presented in our paper are available on the tutorial website
(<https://macsgp-tutorial.readthedocs.io/>).

- [Visium adult mouse brain datasets containing two biological
  replicates](https://macsgp-tutorial.readthedocs.io/en/latest/analysis/mouse_brain/)
- [Visium kidney cancer dataset at the tumour-normal
  interface](https://macsgp-tutorial.readthedocs.io/en/latest/analysis/kidney_cancer/)
- [Multiple human colorectal cancer datasets generated by different
  technologies](https://macsgp-tutorial.readthedocs.io/en/latest/analysis/CRC/)

## Reference

If you find `MacSGP` or any of the source code in this repository useful for your work, please cite:
> Mapping Cell-Type-Specific Spatial Gene Programs Uncovers Tissue Architecture and Microenvironment Organization.  
> Yeqin Zeng, Zhiwei Wang, Yuyao Liu, Yuheng Chen, Jiguang Wang, Hao Chen, and Can Yang.  
> Submitted, 2025.

## Development

The software is developed and maintained by [Yeqin Zeng](mailto:yzengbj@connect.ust.hk).

## Contact

Please feel free to contact [Yeqin Zeng](mailto:yzengbj@connect.ust.hk), [Zhiwei Wang](mailto:zhiwei.wang@connect.ust.hk), or [Prof.Â Can Yang](mailto:macyang@ust.hk) if any inquiries.
